ipld/bitfield/src/lib.rs

```diff
@@ -351,14 +358,46 @@ impl BitOr<&BitField> for &BitField {
 
     #[inline]
     fn bitor(self, rhs: &BitField) -> Self::Output {
-        BitField::from_ranges(self.ranges().union(rhs.ranges()))
+        if self.is_trivially_empty() {
+            rhs.clone()
+        } else if rhs.is_trivially_empty() {
+            self.clone()
+        } else {
+            BitField::from_ranges(self.ranges().union(rhs.ranges()))
+        }
+    }
+}
+
+impl BitOr<BitField> for BitField {
+    type Output = BitField;
+
+    #[inline]
+    fn bitor(self, rhs: BitField) -> Self::Output {
+        if self.is_trivially_empty() {
+            rhs
+        } else if rhs.is_trivially_empty() {
+            self
+        } else {
+            BitField::from_ranges(self.ranges().union(rhs.ranges()))
+        }
     }
 }
 
 impl BitOrAssign<&BitField> for BitField {
     #[inline]
     fn bitor_assign(&mut self, rhs: &BitField) {
-        *self = &*self | rhs;
+        if !rhs.is_trivially_empty() {
+            *self = &*self | rhs;
+        }
+    }
+}
+
+impl BitOrAssign<BitField> for BitField {
+    #[inline]
+    fn bitor_assign(&mut self, rhs: BitField) {
+        // by-value to optimize the case where one is empty.
+        let s = std::mem::take(self);
```

Comment:
also may save a clone

Reply:
I don't quite follow.  The `|` op on 430 uses the `bitor`'s defined above.  Are you saying `impl BitOr<&BitField> for &BitField {` is less efficient than `impl BitOr<BitField> for &BitField {` apart from the clone?  All I can see different is the clone.

Reply:
That's the difference. But I'll see if I can simplify a bit.